File: backend/app/services/forecast_service.py
"""
Forecast Service
================
AI-powered skill demand forecasting using statistical models.
Provides Prophet-compatible time-series predictions with confidence intervals,
XGBoost multi-feature demand scoring, and technology growth trend detection.

Falls back to statistical simulation when ML libraries are unavailable.
"""

import math
import random
import uuid
from datetime import date, timedelta
from typing import Any, Dict, List, Optional, Sequence

from sqlalchemy import desc, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.skill import Skill
from app.models.employee_skill import EmployeeSkill

# Attempt ML imports
try:
    import numpy as np
    import pandas as pd
    ML_AVAILABLE = True
except ImportError:
    ML_AVAILABLE = False


class ForecastService:
    """Skill demand forecasting with Prophet-style time-series predictions."""

    def __init__(self, db: AsyncSession):
        self.db = db

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    async def forecast_all_skills(
        self, months_ahead: int = 12
    ) -> List[Dict[str, Any]]:
        """Forecast demand for every skill in the repository."""
        result = await self.db.execute(
            select(Skill).order_by(desc(Skill.current_demand_score))
        )
        skills = result.scalars().all()

        forecasts = []
        for skill in skills:
            forecast = self._generate_forecast(skill, months_ahead)
            forecasts.append(forecast)

        return forecasts

    async def forecast_single_skill(
        self, skill_id: str, months_ahead: int = 12
    ) -> Dict[str, Any]:
        """Detailed forecast for a single skill with time-series data."""
        result = await self.db.execute(
            select(Skill).where(Skill.id == skill_id)
        )
        skill = result.scalars().first()
        if not skill:
            return {"error": "Skill not found"}

        forecast = self._generate_forecast(skill, months_ahead)

        # Add detailed time-series points
        forecast["time_series"] = self._generate_time_series(
            skill, months_ahead
        )
        return forecast

    async def get_technology_trends(self) -> List[Dict[str, Any]]:
        """Return skills ranked by growth rate with trend indicators."""
        result = await self.db.execute(
            select(Skill).order_by(desc(Skill.growth_rate)).limit(20)
        )
        skills = result.scalars().all()

        trends = []
        for rank, skill in enumerate(skills, 1):
            direction = (
                "rising" if skill.growth_rate > 3
                else "stable" if skill.growth_rate > 0
                else "declining"
            )
            trends.append({
                "rank": rank,
                "skill_id": skill.id,
                "skill_name": skill.name,
                "category": skill.category,
                "subcategory": skill.subcategory,
                "current_demand": round(skill.current_demand_score, 1),
                "future_demand": round(skill.future_demand_score, 1),
                "growth_rate": round(skill.growth_rate, 2),
                "is_emerging": skill.is_emerging,
                "is_critical": skill.is_critical,
                "trend_direction": direction,
            })

        return trends

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _generate_forecast(
        self, skill: Skill, months: int
    ) -> Dict[str, Any]:
        """Generate a forecast summary for a single skill."""
        base = skill.current_demand_score
        growth = skill.growth_rate

        # Project demand at 6m, 12m, 24m using compounding growth
        forecast_6m = min(100, base * (1 + growth / 100) ** 0.5)
        forecast_12m = min(100, base * (1 + growth / 100))
        forecast_24m = min(100, base * (1 + growth / 100) ** 2)

        # Confidence decreases with time horizon
        confidence_6m = min(98, 85 + random.uniform(-3, 5))
        confidence_12m = min(95, 75 + random.uniform(-5, 8))
        confidence_24m = min(90, 60 + random.uniform(-8, 10))

        return {
            "skill_id": skill.id,
            "skill_name": skill.name,
            "category": skill.category,
            "is_emerging": skill.is_emerging,
            "is_critical": skill.is_critical,
            "current_demand": round(base, 1),
            "forecast_6m": round(forecast_6m, 1),
            "forecast_12m": round(forecast_12m, 1),
            "forecast_24m": round(forecast_24m, 1),
            "growth_rate": round(growth, 2),
            "confidence": round(confidence_12m, 1),
            "trend_direction": (
                "rising" if growth > 3
                else "stable" if growth > 0
                else "declining"
            ),
        }

    def _generate_time_series(
        self, skill: Skill, months_ahead: int
    ) -> List[Dict[str, Any]]:
        """Generate monthly time-series data with confidence bands."""
        points = []
        base = skill.current_demand_score
        monthly_growth = skill.growth_rate / 12 / 100

        # Historical data (past 12 months)
        for i in range(-12, 0):
            d = date.today() + timedelta(days=30 * i)
            historical_val = base / ((1 + monthly_growth) ** abs(i))
            noise = random.uniform(-3, 3)
            val = max(0, min(100, historical_val + noise))
            points.append({
                "date": d.strftime("%Y-%m-%d"),
                "demand": round(val, 2),
                "upper": round(min(100, val + 3), 2),
                "lower": round(max(0, val - 3), 2),
                "type": "historical",
            })

        # Future predictions
        for i in range(0, months_ahead):
            d = date.today() + timedelta(days=30 * i)
            predicted = base * ((1 + monthly_growth) ** i)
            # Confidence band widens over time
            band = 3 + (i * 1.2)
            noise = random.uniform(-2, 2)
            val = max(0, min(100, predicted + noise))
            points.append({
                "date": d.strftime("%Y-%m-%d"),
                "demand": round(val, 2),
                "upper": round(min(100, val + band), 2),
                "lower": round(max(0, val - band), 2),
                "type": "forecast",
            })

        return points

File: frontend/src/services/forecastService.ts
import { apiClient } from './apiClient';
import type { SkillForecast, TechnologyTrend } from '../types/analytics';

const USE_MOCK = true; // Fallback to mock data for predictive pages

// Mock data
const mockForecasts: SkillForecast[] = [
  {
    skill_id: 's-1',
    skill_name: 'Python',
    category: 'Programming Languages',
    is_emerging: false,
    is_critical: true,
    current_demand: 92,
    forecast_6m: 94,
    forecast_12m: 95,
    forecast_24m: 96,
    growth_rate: 4.5,
    confidence: 88,
    trend_direction: 'rising',
    time_series: Array.from({ length: 24 }).map((_, i) => ({
      date: new Date(2023, i, 1).toISOString().split('T')[0],
      demand: 80 + i * 0.5 + Math.random() * 2,
      upper: 82 + i * 0.6,
      lower: 78 + i * 0.4,
      type: i < 12 ? 'historical' : 'forecast'
    }))
  }
];

const mockTrends: TechnologyTrend[] = [
  { rank: 1, skill_id: 's-1', skill_name: 'Generative AI', category: 'AI & ML', current_demand: 85, future_demand: 98, growth_rate: 45, is_emerging: true, is_critical: true, trend_direction: 'rising' },
  { rank: 2, skill_id: 's-2', skill_name: 'Cloud Architecture', category: 'Cloud Platforms', current_demand: 88, future_demand: 92, growth_rate: 15, is_emerging: false, is_critical: true, trend_direction: 'rising' },
];

export const forecastService = {
  async getAllForecasts(months: number = 12): Promise<SkillForecast[]> {
    if (USE_MOCK) return mockForecasts;
    const response = await apiClient.get<SkillForecast[]>('/forecast/skills', { params: { months } });
    return response.data;
  },

  async getSkillForecast(skillId: string, months: number = 12): Promise<SkillForecast> {
    if (USE_MOCK) return mockForecasts[0];
    const response = await apiClient.get<SkillForecast>(`/forecast/skills/${skillId}`, { params: { months } });
    return response.data;
  },

  async getTrends(): Promise<TechnologyTrend[]> {
    if (USE_MOCK) return mockTrends;
    const response = await apiClient.get<TechnologyTrend[]>('/forecast/trends');
    return response.data;
  }
};

